torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -671,8 +661,45 @@ def view_avoid_dupes_with_primals(t):
         # This analysis function returns *only* the outputs that are meant to be tangents to the backwards.
         # Anything that aliases (inputs returned in the fw due to metadata mutations, or outputs that alias inputs/intermediates)
         # are *regenerated* later, and not used directly in the autograd graph
+        def _plain_fake_tensor_like_subclass(x):
+            with detect_fake_mode():
+                return torch.empty(
+                    x.shape, dtype=x.dtype, device=x.device, layout=x.layout
+                )
+
         f_input_tangents = [
-            inp
+            # Note: [AOTAutograd Tangent Subclassness for mutated inputs]
+            # Generally when creating tangents to trace with, we assume that tangents will have
+            # the same subclass-ness as their forward outs
+            # however: for tangents that correspond to input mutations, in practice it is more likely
+            # that these tangents will be plain tensors of zeros at runtime, so we tweak our guess
+            # to assume that these tangents should always be plaint tensors.
```

Comment:
Working through this example to try to understand this better:
```
def f(x):
    x.mul_(2)
    return x + 1
```

```
def inner(x):
    x_modified = x * 2
    return x, x_modified, x_modified + 1
    
def outer(x):
    x_input, x_modified, out = inner(x)
    x_input.copy_(x_modified)
    return out
```
Some noob questions (I might be missing some crucial AOTAutograd context):

In this case the tangent that corresponds to the `x` is zeros/undefined right? Is that what you mean by "tangents that correspond to input mutations"? (I was confused by the naming of x_updated_tangent in the comment, because x_updated should have non-zero gradient right)

Is the reason why the tensor is assumed to be a plain tensor because when custom autograd Function receives None, it materializes a zeros in the shape of the output shape and does not reconstruct the subclass? (If this is the case, that's not actually true for NestedTensor which will actually create a subclass in this case).

Reply:
> If this is the case, that's not actually true for NestedTensor which will actually create a subclass in this case

I might be missing something here, because if this is true, then we would've had to use `disable_guess_zero_tangent_for_mutated_input_subclass` to special case NestedTensor...

Reply:
> it materializes a zeros in the shape of the output shape and does not reconstruct the subclass? (If this is the case, that's not actually true for NestedTensor which will actually create a subclass in this case).

oh hmm... @soulitzer where does the logic for that live? (is there some special-casing of NJT in the autograd engine?)

Reply:
Relevant PR here. https://github.com/pytorch/pytorch/pull/136875
maybe relevant issue https://github.com/pytorch/pytorch/issues/133644

Not in engine but in custom autograd Function. For nested tensor specifically it is important that the subclassness of a tensor matches the subclassness of the tangent because the subclassesness is what defines the tensor's shape, and autograd requires the shape to be equivalent.

Reply:
hmm @IvanKobzarev it sounds like we should special-case NJT in that tangent-guessing decision. Any chance you're able to repro the error Jeffrey is talking about? I think you should be able to repro with code like this:
```
@torch.compile
def f(njt):
    njt.mul_(2)
    return njt + 1

# do an input mutation on an NJT that requires grad
x = NestedTensor(..., requires_grad=True)
out = f(x)
# don't use x anywhere after out is computed
# According to Jeffrey, NJT is special-cased so that the `x_updated_tangent` will **not** be a plain ttensor here
out.sum().backward()
```

Reply:
@bdhirsh @soulitzer 
I tried this variant, but for NJT this fails in eager mode, as view of Leaf-variable is mutated:
```
        def fn(x):
            x.mul_(2)
            return x + 1

        def _inp():
            v = torch.zeros(32, 3, requires_grad=True)
            return torch.nested.nested_tensor_from_jagged(
                v,
                offsets=torch.tensor([0, 2, 3]),
            )
        
        ref_x = _inp()
        ref_y = fn(ref_x)
        ref_y.sum().backward()
```
```
Traceback (most recent call last):
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_utils.py", line 3099, in wrapper
    method(*args, **kwargs)
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_utils.py", line 3099, in wrapper
    method(*args, **kwargs)
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_device_type.py", line 480, in instantiated_test
    raise rte
  File "/data/users/ivankobzarev/b/pytorch/torch/testing/_internal/common_device_type.py", line 460, in instantiated_test
    result = test(self, **param_kwargs)
  File "/data/users/ivankobzarev/b/pytorch/test/test_nestedtensor.py", line 8592, in test_nested_tensor_input_mutation_backward
    ref_y = fn(ref_x)
  File "/data/users/ivankobzarev/b/pytorch/test/test_nestedtensor.py", line 8580, in fn
    x.mul_(2)
  File "/data/users/ivankobzarev/b/pytorch/torch/nested/_internal/nested_tensor.py", line 353, in __torch_function__
    return func(*args, **kwargs)
RuntimeError: a view of a leaf Variable that requires grad is being used in an in-place operation.
```

Reply:
From offline discussion:

NJT has a special handling in Autograd, that even if Autograd wants to create zero tensor, for NJT it will call zeros_like() and the tangent is guranteed to be NJT even if it is not used. (see torch/csrc/autograd/python_function.cpp use_zeros_like)

Adding special case to not guess plain_zero_tensor for NJT.

Adding test for this case in test_nestedtensor.py